contracts/Comptroller.sol

```diff
@@ -1239,42 +1182,86 @@ contract Comptroller is ComptrollerV4Storage, ComptrollerInterface, ComptrollerE
-        compSupplierIndex[cToken][supplier] = supplyIndex.mantissa;
-
-        if (supplierIndex.mantissa == 0 && supplyIndex.mantissa > 0) {
-            supplierIndex.mantissa = compInitialIndex;
-        }
-
-        Double memory deltaIndex = sub_(supplyIndex, supplierIndex);
-        uint supplierTokens = CToken(cToken).balanceOf(supplier);
-        uint supplierDelta = mul_(supplierTokens, deltaIndex);
-        uint supplierAccrued = add_(compAccrued[supplier], supplierDelta);
-        compAccrued[supplier] = transferComp(supplier, supplierAccrued, distributeAll ? 0 : compClaimThreshold);
-        emit DistributedSupplierComp(CToken(cToken), supplier, supplierDelta, supplyIndex.mantissa);
+        distributeMarketComp(cToken, supplier, distributeAll, compSupplyState[cToken], compSupplyVestingState[cToken], true, Exp({mantissa: 0}));
     }
 
     /**
      * @notice Calculate COMP accrued by a borrower and possibly transfer it to them
-     * @dev Borrowers will not begin to accrue until after the first interaction with the protocol.
+     * @dev Borrowers will not begin to vest & accrue until after the first interaction with the protocol.
      * @param cToken The market in which the borrower is interacting
      * @param borrower The address of the borrower to distribute COMP to
      */
     function distributeBorrowerComp(address cToken, address borrower, Exp memory marketBorrowIndex, bool distributeAll) internal {
-        CompMarketState storage borrowState = compBorrowState[cToken];
-        Double memory borrowIndex = Double({mantissa: borrowState.index});
-        Double memory borrowerIndex = Double({mantissa: compBorrowerIndex[cToken][borrower]});
-        compBorrowerIndex[cToken][borrower] = borrowIndex.mantissa;
+        distributeMarketComp(cToken, borrower, distributeAll, compBorrowState[cToken], compBorrowVestingState[cToken], false, marketBorrowIndex);
+    }
+
+    /**
+     * @notice Calculate COMP accrued by a holder (supplier or borrower) and possibly transfer it to them
+     * @param cToken The market in which the holder is interacting
+     * @param holder The address of the holder to distribute COMP to
+     */
+    function distributeMarketComp(address cToken, address holder, bool distributeAll, CompMarketState storage marketState, CompMarketState storage vestingState, bool isSupply,
+            Exp memory marketBorrowIndex) internal {
+        Double memory marketIndex = Double({mantissa: marketState.index});
+        Double memory holderIndex;
+        if (isSupply) {
+            holderIndex = Double({mantissa: compSupplierIndex[cToken][holder]});
+            compSupplierIndex[cToken][holder] = marketIndex.mantissa;
+        } else {
+            holderIndex = Double({mantissa: compBorrowerIndex[cToken][holder]});
+            compBorrowerIndex[cToken][holder] = marketIndex.mantissa;
+        }
+
+        if (isSupply) {
```

Comment:
should these lines go in the previous isSupply clause?

Reply:
Thanks, fixed!